core/services/relay/evm/binding/named.go

```diff
@@ -0,0 +1,263 @@
+			}
+
+			rdr, exists := cb.GetReaderNamed(values.method)
+			if !exists {
+				continue
+			}
+
+			// TODO: need address for batch call
+			call, err := rdr.BatchCall(common.HexToAddress(values.address), req.Params, req.ReturnVal)
+			if err != nil {
+				return nil, err
+			}
+
+			batchCall = append(batchCall, call)
+		}
+	}
+
+	results, err := b.batch.BatchCall(ctx, 0, batchCall)
+	if err != nil {
+		return nil, err
+	}
+
+	// reconstruct results from batchCall and filteredLogs into common type while maintaining order from request.
+	batchGetLatestValuesResults := make(commontypes.BatchGetLatestValuesResult)
+	for contractName, contractResult := range results {
+		batchGetLatestValuesResults[contractName] = commontypes.ContractBatchResults{}
+		for _, methodResult := range contractResult {
+			brr := commontypes.BatchReadResult{
+				ReadIdentifier: types.BoundContract{
+					Address: methodResult.Address,
+					Name:    contractName,
+				}.ReadIdentifier(methodResult.MethodName),
+			}
+
+			brr.SetResult(methodResult.ReturnValue, methodResult.Err)
+			batchGetLatestValuesResults[contractName] = append(batchGetLatestValuesResults[contractName], brr)
+		}
+	}
+
+	return batchGetLatestValuesResults, err
+}
+
+// Unbind unbinds contract addresses to contract bindings and read bindings.
+func (b *NamedBindings) Unbind(ctx context.Context, reg Registrar, bindings []commontypes.BoundContract) error {
+	b.mu.RLock()
+	defer b.mu.RUnlock()
+
+	for _, binding := range bindings {
+		contract, exists := b.contractBindings[binding.Name]
+		if !exists {
+			return fmt.Errorf("%w: no contract named %s", commontypes.ErrInvalidConfig, binding.Name)
+		}
+
+		b.contractLookup.unbindAddressForContract(binding.Name, binding.Address)
+
+		if err := errors.Join(
+			contract.Unbind(ctx, reg, common.HexToAddress(binding.Address)),
+			contract.UnbindReaders(ctx, common.HexToAddress(binding.Address)),
+		); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func (b *NamedBindings) RegisterAll(ctx context.Context, reg Registrar) error {
+	b.mu.RLock()
+	defer b.mu.RUnlock()
+
+	for _, cb := range b.contractBindings {
+		if err := errors.Join(cb.RegisterReaders(ctx), cb.Register(ctx, reg)); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func (b *NamedBindings) UnregisterAll(ctx context.Context, reg Registrar) error {
+	b.mu.RLock()
+	defer b.mu.RUnlock()
+
+	for _, cb := range b.contractBindings {
+		if err := errors.Join(cb.UnregisterReaders(ctx), cb.Unregister(ctx, reg)); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func (b *NamedBindings) SetCodecAll(codec commontypes.RemoteCodec) {
+	b.mu.RLock()
+	defer b.mu.RUnlock()
+
+	for _, cb := range b.contractBindings {
+		cb.SetCodecAll(codec)
+	}
+}
+
+func (b *NamedBindings) WithFilter(name string, filter logpoller.Filter) {
+	b.mu.RLock()
+	defer b.mu.RUnlock()
+
+	if contract, ok := b.contractBindings[name]; ok {
```

Comment:
This should throw an error, if we get to this point something went very wrong because we either have a contract without methods or events or a disappearing map entry